File: 프로그래머스/unrated/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.py
def solution(bandage, health, attacks):
    # 붕대 감기는 t초동안 x/s 체력을 회복함.
    # t초 연속으로 붕대 감으면 y만큼 추가 회복
    # 최대 체력 초과는 안됨.
    # 공격당하면 기술 취소됨. (공격 당하는 순간은 회복x)
    # 붕대 감기 취소되면 바로 다시 사용함. 성공 시간 0초로 초기화
    # 체력 0 이하되면 사망
    # bandage = [시전 시간, 초당 회복량, 추가 회복량]
    # attacks[i] = [공격 시간, 피해량] # 공격 시간을 기준으로 오름차순 정렬됨.
    
    # attacks를 for문 돌리기.
    hp = health
    for i in range(len(attacks)):
        hp -= attacks[i][1]
        if(hp <= 0):
            return -1
        if(i < len(attacks) - 1):
            # 붕대를 감을 시간은 이번 공격 시간부터 과 다음 공격 직전 시간까지
            time = attacks[i + 1][0] - attacks[i][0] - 1
            # 붕대 감은 초 만큼 회복, 1회 완료마다 추가 회복
            healing = (time * bandage[1]) + (time//bandage[0] * bandage[2])
            hp = (hp + healing) if((hp + healing) < health) else health
        print(i, hp)
    return hp

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.java
class Solution {
    
    public int solution(int[] bandage, int health, int[][] attacks) {
//         붕대는 t초 동안 감으면서, 1초에 x만큼의 체력을 회복함.
//         끊기지 않으면 y 만큼 추가 회복함.
//         최대 체력 이상은 불가능.
//         공격 당하면 기술 취소되고, 그 순산에는 체력 회복 x
//         붕대감기는 무한 반복됨.
//         체력이 0 이하가 되면 사망.
        
//         모든 공격 직후 체력 or 죽으면 -1 리턴
        
        // 풀피일 때 연속 회복 성공 시간은 필요 없음 -> 어차피 공격당하면 회복 x, 초기화
        
        int answer = health;
        for(int i = 0; i < attacks.length; i++){
            answer -= attacks[i][1];
            
            // 피 0 이하면 사망
            if(answer <= 0) return -1;
            
            // 마지막이 아니면 힐
            if(i != attacks.length - 1){
                int time = attacks[i + 1][0] - attacks[i][0] - 1;

                // 피 계산
                answer += bandage[1] * time + time/bandage[0] * bandage[2];
                if(answer > health) answer = health;                
            }
        }
        return answer;
    }
}